README.md
Solr River Plugin for ElasticSearch [![Build Status](https://buildhive.cloudbees.com/job/javanna/job/elasticsearch-river-solr/badge/icon)](https://buildhive.cloudbees.com/job/javanna/job/elasticsearch-river-solr/) [![Build Status](https://travis-ci.org/javanna/elasticsearch-river-solr.png)](https://travis-ci.org/javanna/elasticsearch-river-solr) [![Build Status](https://drone.io/github.com/javanna/elasticsearch-river-solr/status.png)](https://drone.io/github.com/javanna/elasticsearch-river-solr/latest)
==================================

The Solr River plugin allows to import data from [Apache Solr](http://lucene.apache.org/solr) to [elasticsearch](http://www.elasticsearch.org).

In order to install the latest version of the plugin, simply run: `bin/plugin install river-solr -url http://bit.ly/1grj1ny`.
You can copy paste the url of a specific version from the table below, depending on the elasticsearch version you're running.


Versions
--------

<table>
	<thead>
		<tr>
			<td>Solr River Plugin</td>
			<td>Elasticsearch</td>
		</tr>
	</thead>
	<tbody>
	    <tr>
            <td>master</td>
            <td>1.0.x</td>
        </tr>
        <tr>
            <td><a href ="http://bit.ly/1grj1ny">2.0</a></td>
            <td>1.0.x</td>
        </tr>
        <tr>
            <td><a href="http://bit.ly/1exH5mG">1.1</a></td>
            <td>0.90.x</td>
        </tr>
        <tr>
            <td><a href="http://bit.ly/10ioSk0">1.0.4</a></td>
            <td>0.90.0</td>
        </tr>
	    <tr>
    	    <td><a href="http://bit.ly/12rmrSN">1.0.3</a></td>
    		<td>0.20.0 -> 0.20.6</td>
        </tr>
		<tr>
			<td><a href="http://bit.ly/15cCMIB">1.0.2</a></td>
			<td>0.20.0 -> 0.20.6</td>
		</tr>
		<tr>
            <td><a href="http://bit.ly/Yo61UW">1.0.1</a></td>
            <td>0.19.3 -> 0.19.12</td>
        </tr>
		<tr>
            <td><a href="http://bit.ly/XUl2LZ">1.0.0</a></td>
        	<td>0.19.3 -> 0.19.12</td>
        </tr>
	</tbody>
</table>


Getting Started
===============

The Solr River allows to query a running Solr instance and index the returned documents in elasticsearch.
It retrieves documents via [json response writer](http://wiki.apache.org/solr/SolJSON), through http get requests (solrj is not used anymore to communicate with Solr).

All the [common query parameters](http://wiki.apache.org/solr/CommonQueryParameters) are supported.

The solr river is not meant to keep solr and elasticsearch in sync, that's why it automatically deletes itself on completion, so that the river doesn't start up again at every node restart.
This is the default behaviour, which can be disabled through the `close_on_completion` parameter.


Installation
------------

Here is how you can easily create the river and index data from Solr, just providing the solr url and the query to execute:

```sh
curl -XPUT localhost:9200/_river/solr_river/_meta -d '
{
    "type" : "solr",
    "solr" : {
        "url" : "http://localhost:8080/solr/",
        "q" : "*:*"
    }
}'
```

All supported parameters are optional. The following example request contains all the parameters that are supported together with the corresponding default values applied when not present.

```javascript
{
    "type" : "solr",
    "close_on_completion" : "true",
    "solr" : {
        "url" : "http://localhost:8983/solr/",
        "q" : "*:*",
        "fq" : "",
        "fl" : ""
        "qt" : "",
        "uniqueKey" : "id",
        "rows" : 10
    },
    "index" : {
        "index" : "solr",
        "type" : "import",
        "bulk_size" : 100,
        "max_concurrent_bulk" : 10,
        "mapping" : "",
        "settings": ""
    }
}
```

The fq and fl parameters can be provided as either an array or a single value.

You can provide your own mapping while creating the river, as well as the index settings, which will be used when creating the new index if needed.

The index is created when not already existing, otherwise the documents are added to the existing one with the configured name.

The documents are indexed using the [bulk api](http://www.elasticsearch.org/guide/reference/java-api/bulk.html).
You can control the size of each bulk (default 100) and the maximum number of concurrent bulk operations (default is 10).
Once the limit is reached the indexing will slow down, waiting for one of the bulk operations to finish its work; no documents will be lost.

Transform documents
------------
Since version 1.0.3 it's possible to transform the documents via scripting. The feature works exactly as the [update api](http://www.elasticsearch.org/guide/reference/api/update.html). The needed parameters can be specified within the transform section while registering the river, like this:

```javascript
{
    "type" : "solr",
    "solr" : {
        "url" : "http://localhost:8983/solr/",
        "q" : "*:*",
    },
    "index" : {
        "index" : "solr",
        "type" : "import",
    },
    "transform" : {
        "script" : "ctx._source.counter += count",
        "params" : {
            "count" : 4
        }
    }
}
```

The example above increments by 4 the content of the counter field for every document right before the indexing process in elasticsearch.

Limitations
------------

* only stored fields can be retrieved from Solr, therefore indexed in elasticsearch
* the river is not meant to keep elasticsearch in sync with Solr, but only to import data once. It's possible to register the river multiple times in order to import different sets of documents though, even from different solr instances.
* it's recommended to create the [mapping](http://www.elasticsearch.org/guide/reference/mapping/index.html) given the existing solr schema in order to apply the correct text analysis while importing the documents. In the future there might be an option to auto generating it from the Solr schema.

License
=======

```
This software is licensed under the Apache 2 license, quoted below.

Copyright 2014 Luca Cavanna

Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
```


CHANGES.txt
Version 2.0
Date: 16/02/2014
Changes:
* Updated to elasticsearch 1.0.0

Version 1.1
Date: 26/11/2013
Changes:
* Updated to elasticsearch 0.90.7
* Issue #6: Removed solrj dependency

Version 1.0.4
Date: 06/05/2013
Changes:
* Updated to elasticsearch 0.90.0

Version: 1.0.3
Date: 05/03/2013
Changes:
* Issue #3: Added support for transforming documents via scripting

Version: 1.0.2
Date: 24/02/2013
Changes:
* Issue #2: Remove bulk custom code and use BulkProcessor. Supported since elasticsearch 0.19.9

Version: 1.0.1
Date: 21/11/2012
Changes:
* Issue #1: Added close_on_completion option (default true) to automatically delete the river at the end of the data import process.

Version: 1.0.0
Date: 12/11/2012
* First release

src/test/java/org/elasticsearch/river/solr/support/Field.java
/*
 * Licensed to Luca Cavanna (the "Author") under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. Elastic Search licenses this
 * file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.elasticsearch.river.solr.support;

public class Field<Value> {

    private final String name;
    private final Value value;

    public Field(String name, Value value) {
        this.name = name;
        this.value = value;
    }

    public Value getValue() {
        return value;
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return "Field{" +
                "name='" + name + '\'' +
                ", value=" + value +
                '}';
    }
}


src/test/java/org/elasticsearch/river/solr/support/Random.java
/*
 * Licensed to Luca Cavanna (the "Author") under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. Elastic Search licenses this
 * file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.elasticsearch.river.solr.support;

import org.apache.commons.lang.RandomStringUtils;

import java.util.Calendar;
import java.util.Date;

public final class Random {

    private static final java.util.Random RANDOM = new java.util.Random();

    private Random() {

    }

    public static String nextSentence(int numberOfWords) {
        //0 to numberOfWords number of words for each sentence
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < RANDOM.nextInt(numberOfWords); i++) {
            builder.append(nextWord()).append(" ");
        }
        return builder.toString();
    }

    public static String nextWord() {
        //1 to 10 number of characters for each random word
        return RandomStringUtils.randomAlphabetic(RANDOM.nextInt(9)+1);
    }

    public static Date nextDate() {
        int month = RANDOM.nextInt(11) + 1;
        int year = RANDOM.nextInt(30) + 1980;
        int date = RANDOM.nextInt(365);
        Calendar calendar = Calendar.getInstance();
        calendar.set(year, month, date);
        return calendar.getTime();
    }

    public static int nextInt(int n) {
        return RANDOM.nextInt(n);
    }

    public static boolean nextBoolean() {
        return RANDOM.nextBoolean();
    }
}


src/test/java/org/elasticsearch/river/solr/support/SolrIndexer.java
/*
 * Licensed to Luca Cavanna (the "Author") under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. Elastic Search licenses this
 * file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.elasticsearch.river.solr.support;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

import java.io.IOException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Map;
import java.util.TimeZone;

public class SolrIndexer {

    private final CloseableHttpClient httpClient;
    private final String solrUrl;

    private final ObjectMapper objectMapper;

    public SolrIndexer(String solrUrl) {
        this.httpClient = HttpClients.createDefault();
        this.solrUrl = solrUrl + "/update";
        this.objectMapper = new ObjectMapper();
        this.objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
        dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
        objectMapper.setDateFormat(dateFormat);
    }

    public void indexDocuments(Map<String, Map<String, Object>> documents) throws IOException {
        StringBuilder jsonBuilder = new StringBuilder("[");
        int i=0;
        for (Map<String, Object> doc : documents.values()) {
            jsonBuilder.append(objectMapper.writeValueAsString(doc));

            if (i < documents.values().size() - 1) {
                jsonBuilder.append(",");
            }
            i++;
        }
        jsonBuilder.append("]");

        HttpPost httpPost = new HttpPost(solrUrl + "?commit=true");
        httpPost.setHeader("Content-type", "application/json");
        httpPost.setEntity(new StringEntity(jsonBuilder.toString()));

        CloseableHttpResponse response = httpClient.execute(httpPost);
        try {
            EntityUtils.consume(response.getEntity());
            if (response.getStatusLine().getStatusCode() != 200) {
                throw new RuntimeException("documents were not properly indexed");
            }
        } finally {
            EntityUtils.consume(response.getEntity());
            response.close();
        }
    }

    public void clearDocuments() throws IOException {
        HttpPost httpPost = new HttpPost(solrUrl + "?commit=true");
        httpPost.setHeader("Content-type", "application/xml");
        httpPost.setEntity(new StringEntity("<delete><query>*:*</query></delete>"));

        CloseableHttpResponse response = httpClient.execute(httpPost);
        try {
            if (response.getStatusLine().getStatusCode() != 200) {
                throw new RuntimeException("documents were not properly cleared");
            }
        } finally {
            EntityUtils.consume(response.getEntity());
            response.close();
        }
    }
}


src/main/java/org/elasticsearch/river/solr/SolrRiverModule.java
/*
 * Licensed to Luca Cavanna (the "Author") under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. Elastic Search licenses this
 * file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.elasticsearch.river.solr;

import org.elasticsearch.common.inject.AbstractModule;
import org.elasticsearch.river.River;

/**
 * Module for the solr river
 */
public class SolrRiverModule extends AbstractModule {

    @Override
    protected void configure() {
        bind(River.class).to(SolrRiver.class).asEagerSingleton();
    }
}


src/main/java/org/elasticsearch/plugin/river/solr/SolrRiverPlugin.java
/*
 * Licensed to Luca Cavanna (the "Author") under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. Elastic Search licenses this
 * file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.elasticsearch.plugin.river.solr;

import org.elasticsearch.common.inject.Inject;
import org.elasticsearch.plugins.AbstractPlugin;
import org.elasticsearch.river.RiversModule;
import org.elasticsearch.river.solr.SolrRiverModule;

/**
 * Plugin class for the Solr river
 */
public class SolrRiverPlugin extends AbstractPlugin {

    @Inject
    public SolrRiverPlugin() {
    }

    @Override
    public String name() {
        return "river-solr";
    }

    @Override
    public String description() {
        return "River Solr plugin";
    }

    /**
     * Registers the {@link SolrRiverModule}
     * @param module the elasticsearch module used to handle rivers
     */
    public void onModule(RiversModule module) {
        module.registerRiver("solr", SolrRiverModule.class);
    }
}


src/main/java/org/elasticsearch/river/solr/SolrRiver.java
/*
 * Licensed to Luca Cavanna (the "Author") under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. Elastic Search licenses this
 * file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.elasticsearch.river.solr;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;
import org.elasticsearch.ElasticsearchIllegalArgumentException;
import org.elasticsearch.action.admin.indices.create.CreateIndexRequestBuilder;
import org.elasticsearch.action.bulk.BulkProcessor;
import org.elasticsearch.action.bulk.BulkRequest;
import org.elasticsearch.action.bulk.BulkResponse;
import org.elasticsearch.action.index.IndexRequest;
import org.elasticsearch.client.Client;
import org.elasticsearch.client.Requests;
import org.elasticsearch.common.Strings;
import org.elasticsearch.common.collect.Maps;
import org.elasticsearch.common.collect.Tuple;
import org.elasticsearch.common.inject.Inject;
import org.elasticsearch.common.xcontent.XContentHelper;
import org.elasticsearch.common.xcontent.XContentType;
import org.elasticsearch.common.xcontent.support.XContentMapValues;
import org.elasticsearch.river.AbstractRiverComponent;
import org.elasticsearch.river.River;
import org.elasticsearch.river.RiverName;
import org.elasticsearch.river.RiverSettings;
import org.elasticsearch.script.ExecutableScript;
import org.elasticsearch.script.ScriptService;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.charset.Charset;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Solr river which allows to index data taken from a running Solr instance
 */
public class SolrRiver extends AbstractRiverComponent implements River {

    private final Client client;
    private final ScriptService scriptService;

    private final String solrUrl;

    private final String query;
    private final String[] filterQueries;
    private final String[] fields;
    private final String requestHandler;
    private final String uniqueKey;
    private final int rows;

    private final String indexName;
    private final String typeName;
    private final String settings;
    private final String mapping;
    private final boolean closeOnCompletion;

    private final String script;
    private final Map<String, Object> scriptParams;
    private final String scriptLang;

    private volatile BulkProcessor bulkProcessor;
    private AtomicInteger start = new AtomicInteger(0);

    private final ObjectMapper objectMapper;
    private final CloseableHttpClient httpClient;

    static final String DEFAULT_UNIQUE_KEY = "id";

    @Inject
    @SuppressWarnings("unchecked")
    protected SolrRiver(RiverName riverName, RiverSettings riverSettings, Client client, ScriptService scriptService) {
        super(riverName, riverSettings);
        this.client = client;
        this.scriptService = scriptService;

        this.closeOnCompletion = XContentMapValues.nodeBooleanValue(riverSettings.settings().get("close_on_completion"), true);

        String url = "http://localhost:8983/solr/";
        String q = "*:*";
        String uniqueKey = DEFAULT_UNIQUE_KEY;
        int rows = 10;
        String qt = "select";
        String[] fq, fl;
        fq = fl = null;
        if (riverSettings.settings().containsKey("solr")) {
            Map<String, Object> solrSettings = (Map<String, Object>) riverSettings.settings().get("solr");
            url = XContentMapValues.nodeStringValue(solrSettings.get("url"), url);
            q = XContentMapValues.nodeStringValue(solrSettings.get("q"), q);
            rows = XContentMapValues.nodeIntegerValue(solrSettings.get("rows"), rows);
            fq = readArrayOrString(solrSettings.get("fq"));
            fl = readArrayOrString(solrSettings.get("fl"));
            qt = XContentMapValues.nodeStringValue(solrSettings.get("qt"), qt);
            uniqueKey = XContentMapValues.nodeStringValue(solrSettings.get("uniqueKey"), uniqueKey);
        }
        this.solrUrl = url;
        this.query = q;
        this.rows = rows;
        this.uniqueKey = uniqueKey;
        this.filterQueries = fq;
        this.fields = fl;
        this.requestHandler = qt;

        String index = riverName.type();
        String type = "import";
        int maxConcurrentBulk = 10;
        int bulkSize = 100;
        String mapping = null;
        String settings = null;
        if (riverSettings.settings().containsKey("index")) {
            Map<String, Object> indexSettings = (Map<String, Object>) riverSettings.settings().get("index");
            index = XContentMapValues.nodeStringValue(indexSettings.get("index"), index);
            type = XContentMapValues.nodeStringValue(indexSettings.get("type"), type);
            bulkSize = XContentMapValues.nodeIntegerValue(indexSettings.get("bulk_size"), bulkSize);
            maxConcurrentBulk = XContentMapValues.nodeIntegerValue(indexSettings.get("max_concurrent_bulk"), maxConcurrentBulk);
            settings = XContentMapValues.nodeStringValue(indexSettings.get("settings"), settings);
            mapping = XContentMapValues.nodeStringValue(indexSettings.get("mapping"), mapping);
        }
        this.settings = settings;
        this.mapping = mapping;
        this.indexName = index;
        this.typeName = type;

        String script = null;
        Map<String, Object> scriptParams = Maps.newHashMap();
        String scriptLang = null;
        if (riverSettings.settings().containsKey("transform")) {
            Map<String, Object> transformSettings = (Map<String, Object>) riverSettings.settings().get("transform");
            script = XContentMapValues.nodeStringValue(transformSettings.get("script"), null);
            scriptLang = XContentMapValues.nodeStringValue(transformSettings.get("lang"), null);
            scriptParams = (Map<String, Object>) transformSettings.get("params");
        }
        this.script = script;
        this.scriptParams = scriptParams;
        this.scriptLang = scriptLang;


        this.bulkProcessor = BulkProcessor.builder(client, new BulkProcessor.Listener() {
            @Override
            public void beforeBulk(long executionId, BulkRequest request) {
                logger.info("Going to execute new bulk composed of {} actions", request.numberOfActions());
            }

            @Override
            public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
                logger.info("Executed bulk composed of {} actions", request.numberOfActions());
            }

            @Override
            public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
                logger.warn("Error executing bulk", failure);
            }
        }).setBulkActions(bulkSize).setConcurrentRequests(maxConcurrentBulk).build();

        this.objectMapper = new ObjectMapper();
        this.objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);

        this.httpClient = HttpClients.createDefault();
    }

    private String[] readArrayOrString(Object node) {
        if (XContentMapValues.isArray(node)) {
            List list = (List) node;
            String[] array = new String[list.size()];
            for (int i = 0; i < list.size(); i++) {
                array[i] = XContentMapValues.nodeStringValue(list.get(i), null);
            }
            return array;
        }

        String value = XContentMapValues.nodeStringValue(node, null);
        if (value != null) {
            return new String[]{value};
        }
        return null;
    }


    @Override
    public void start() {

        if (!client.admin().indices().prepareExists(indexName).execute().actionGet().isExists()) {

            CreateIndexRequestBuilder createIndexRequest = client.admin().indices()
                    .prepareCreate(indexName);

            if (settings != null) {
                createIndexRequest.setSettings(settings);
            }
            if (mapping != null) {
                createIndexRequest.addMapping(typeName, mapping);
            }

            createIndexRequest.execute().actionGet();
        }

        StringBuilder baseSolrQuery = createSolrQuery();

        Long numFound = null;
        int startParam;
        while ((startParam = start.getAndAdd(rows)) == 0 || startParam < numFound) {
            String solrQuery = baseSolrQuery.toString() + "&start=" + startParam;
            CloseableHttpResponse httpResponse = null;
            try {
                logger.info("Sending query to Solr: {}", solrQuery);
                httpResponse = httpClient.execute(new HttpGet(solrQuery));

                if (httpResponse.getStatusLine().getStatusCode() != 200) {
                    logger.error("Solr returned non ok status code: {}", httpResponse.getStatusLine().getReasonPhrase());
                    EntityUtils.consume(httpResponse.getEntity());
                    continue;
                }

                JsonNode jsonNode = objectMapper.readTree(EntityUtils.toString(httpResponse.getEntity()));
                JsonNode response = jsonNode.get("response");
                JsonNode numFoundNode = response.get("numFound");
                numFound = numFoundNode.asLong();
                if (logger.isWarnEnabled() && numFound == 0) {
                    logger.warn("The solr query {} returned 0 documents", solrQuery);
                }

                Iterator<JsonNode> docsIterator = response.get("docs").iterator();

                while (docsIterator.hasNext()) {
                    JsonNode docNode = docsIterator.next();

                    try {
                        JsonNode uniqueKeyNode = docNode.get(uniqueKey);

                        if (uniqueKeyNode == null) {
                            logger.error("The uniqueKey value is null");
                        } else {
                            String id = uniqueKeyNode.asText();
                            ((ObjectNode) docNode).remove(uniqueKey);

                            IndexRequest indexRequest = Requests.indexRequest(indexName).type(typeName).id(id);
                            String source = objectMapper.writeValueAsString(docNode);
                            if (this.script == null) {
                                indexRequest.source(source);
                            } else {
                                Tuple<XContentType, Map<String, Object>> newSourceAndContent = transformDocument(source);
                                indexRequest.source(newSourceAndContent.v2(), newSourceAndContent.v1());
                            }

                            bulkProcessor.add(indexRequest);

                        }
                    } catch (IOException e) {
                        logger.warn("Error while importing documents from solr to elasticsearch", e);
                    }
                }
            } catch (IOException e) {
                logger.error("Error while executing the solr query [" + solrQuery + "]", e);
                bulkProcessor.close();
                try {
                    httpClient.close();
                } catch(IOException ioe) {
                    logger.warn(e.getMessage(), ioe);
                }
                //if a query fails the next ones are most likely going to fail too
                return;
            } finally {
                if (httpResponse != null) {
                    try {
                        httpResponse.close();
                    } catch (IOException e) {
                        logger.error(e.getMessage(), e);
                    }
                }
            }
        }

        bulkProcessor.close();
        try {
            httpClient.close();
        } catch (IOException e) {
            logger.warn(e.getMessage(), e);
        }

        logger.info("Data import from solr to elasticsearch completed");

        if (closeOnCompletion) {
            logger.info("Deleting river");
            client.admin().indices().prepareDeleteMapping("_river").setType(riverName.name()).execute();
        }
    }

    @SuppressWarnings("unchecked")
    protected Tuple<XContentType, Map<String, Object>> transformDocument(String jsonDocument) {
        Tuple<XContentType, Map<String, Object>> sourceAndContent = XContentHelper.convertToMap(jsonDocument.getBytes(Charset.forName("utf-8")), true);

        Map<String, Object> ctx = new HashMap<String, Object>(2);
        ctx.put("_source", sourceAndContent.v2());

        try {
            ExecutableScript executableScript = scriptService.executable(scriptLang, script, scriptParams);
            executableScript.setNextVar("ctx", ctx);
            executableScript.run();
            ctx = (Map<String, Object>) executableScript.unwrap(ctx);
        } catch (Exception e) {
            throw new ElasticsearchIllegalArgumentException("failed to execute script", e);
        }

        final Map<String, Object> updatedSourceAsMap = (Map<String, Object>) ctx.get("_source");
        return new Tuple<XContentType, Map<String, Object>>(sourceAndContent.v1(), updatedSourceAsMap);
    }

    protected StringBuilder createSolrQuery()  {
        StringBuilder queryBuilder = new StringBuilder(solrUrl);

        if (Strings.hasLength(requestHandler)) {
            if (queryBuilder.charAt(queryBuilder.length() - 1) != '/') {
                queryBuilder.append("/");
            }
            queryBuilder.append(requestHandler);
        }

        queryBuilder.append("?q=").append(encode(query)).append("&wt=json");
        if (filterQueries != null) {
            for (String filterQuery : filterQueries) {
                queryBuilder.append("&fq=").append(encode(filterQuery));
            }
        }
        if (fields != null) {
            queryBuilder.append("&fl=");
            for (int i = 0; i < fields.length; i++) {
                if (i>0) {
                    queryBuilder.append(encode(" "));
                }
                queryBuilder.append(encode(fields[i]));
            }
        }

        queryBuilder.append("&rows=").append(rows);
        return queryBuilder;
    }

    private static String encode(String value) {
        try {
            return URLEncoder.encode(value, "utf-8");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void close() {

    }
}
